File: sample-chat-obj-c/sample-chat/Components/ChatScreen/MediaScreen/ZoomedAttachment screen/ZoomedAttachmentViewController.m
//
//  ZoomedAttachmentViewController.m
//  sample-chat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "ZoomedAttachmentViewController.h"
#import <Photos/Photos.h>
#import "UIViewController+Alert.h"

@interface ZoomedAttachmentViewController ()
@property (strong, nonatomic) UIBarButtonItem *infoItem;
@end

@implementation ZoomedAttachmentViewController

#pragma mark - Life Cycle
- (instancetype)initWithImage:(UIImage *)image
{
    self = [super init];
    if (self) {
        self.zoomImageView = [[UIImageView alloc] initWithImage:image];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIBarButtonItem *backButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"chevron"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(didTapBack:)];
    
    self.navigationItem.leftBarButtonItem = backButtonItem;
    backButtonItem.tintColor = UIColor.whiteColor;
    
    __weak __typeof(self) weakSelf = self;
    UIAction *saveImageAction = [UIAction actionWithTitle:@"Save attachment" image:nil identifier:nil handler:^(__kindof UIAction * _Nonnull action) {
        __typeof(weakSelf)strongSelf = weakSelf;
        UIImage *image = strongSelf.zoomImageView.image;
        if (!image) {
            return;
        }
        [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
            switch (status) {
                case PHAuthorizationStatusAuthorized:
                    UIImageWriteToSavedPhotosAlbum(image, strongSelf, @selector(image:didFinishSavingWithError:contextInfo:), nil);
                    break;
                    
                case PHAuthorizationStatusDenied:
                    NSLog(@"User denied");
                    break;
                    
                default:
                    NSLog(@"Restricted");
                    break;
            }
        }];
    }];
    UIMenu *menu = [UIMenu menuWithTitle:@"" children: @[saveImageAction]];
    self.infoItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"moreInfo"] menu:menu];
    self.navigationItem.rightBarButtonItem = self.infoItem;
    self.infoItem.tintColor = UIColor.whiteColor;
    
    self.navigationController.navigationBar.barTintColor = [UIColor blackColor];
    self.navigationController.navigationBar.barStyle = UIBarStyleBlack;
    self.navigationController.navigationBar.shadowImage = UIImage.new;
    [self.navigationController.navigationBar setTranslucent:NO];
    self.navigationController.navigationBar.tintColor = UIColor.whiteColor;
    self.navigationController.navigationBar.titleTextAttributes = @{ NSForegroundColorAttributeName:UIColor.whiteColor};
    
    self.view.backgroundColor = [UIColor blackColor];
    self.zoomImageView.frame = self.view.frame;
    self.zoomImageView.contentMode = UIViewContentModeScaleAspectFit;
    [self.view addSubview:self.zoomImageView];
    CGFloat topBarHeight = self.view.window.windowScene.statusBarManager.statusBarFrame.size.height + self.navigationController.navigationBar.frame.size.height;
    self.zoomImageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.zoomImageView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.zoomImageView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.zoomImageView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.zoomImageView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-topBarHeight].active = YES;
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didTapBackAction:)];
    [self.view addGestureRecognizer:tapGesture];
}

#pragma mark - Actions
- (void)didTapBack:(UIButton *)sender {
    [self dismissViewControllerAnimated:NO completion:nil];
}

- (void)didTapBackAction:(UITapGestureRecognizer*)sender {
    [self dismissViewControllerAnimated:NO completion:nil];
}

- (void)image:(UIImage *)image didFinishSavingWithError:(NSError *)error contextInfo:(void *)contextInfo {
    NSString *errorMessage = error ? @"Save error" : @"Saved!";
    [self showAnimatedAlertWithTitle:nil message:errorMessage fromViewController:self];
}

@end

File: sample-chat-swift/sample-chat-swift/Components/ChatScreen/MediaScreen/ZoomedAttachmentViewController.swift
//
//  ZoomedAttachmentViewController.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/28/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit
import Photos

class ZoomedAttachmentViewController: UIViewController {
    //MARK: - Properties
    let zoomImageView = UIImageView()
    
    //MARK: - Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let backButtonItem = UIBarButtonItem(image: UIImage(named: "chevron"),
                                             style: .plain,
                                             target: self,
                                             action: #selector(didTapBack(_:)))
        navigationItem.leftBarButtonItem = backButtonItem
        backButtonItem.tintColor = .white
        
        let saveAttachmentAction = UIAction(title: "Save attachment") { [weak self]  action in
            guard let self = self else { return }
            guard let image = self.zoomImageView.image else {
                return
            }
            PHPhotoLibrary.requestAuthorization
                { [weak self] (status) -> Void in
                    switch (status)
                    {
                    case .authorized:
                        UIImageWriteToSavedPhotosAlbum(image, self, #selector(self?.image(_:didFinishSavingWithError:contextInfo:)), nil)
                    case .denied:
                        // Permission Denied
                        debugPrint("[ZoomedAttachmentViewController] User denied")
                    default:
                        debugPrint("[ZoomedAttachmentViewController] Restricted")
                    }
            }
        }
        let menu = UIMenu(title: "", children: [saveAttachmentAction])
        let infoItem = UIBarButtonItem(image: UIImage(named: "moreInfo"),
                                                    menu: menu)
        navigationItem.rightBarButtonItem = infoItem
        infoItem.tintColor = .white
        
        view.backgroundColor = .black
        zoomImageView.contentMode = .scaleAspectFit
        view.addSubview(zoomImageView)
        let navigationBarHeight = self.navigationController?.navigationBar.frame.size.height ?? 44.0
        zoomImageView.translatesAutoresizingMaskIntoConstraints = false
        zoomImageView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        zoomImageView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        zoomImageView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        zoomImageView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -(navigationBarHeight)).isActive = true
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(goBackAction(_ :)))
        view.addGestureRecognizer(tapGesture)
    }
    
    //MARK: - Actions
    @objc private func goBackAction(_ sender: UITapGestureRecognizer) {
        dismiss(animated: false, completion: nil)
    }
    
    @objc func didTapBack(_ sender: UIBarButtonItem) {
        dismiss(animated: false, completion: nil)
    }
    
    @objc func image(_ image: UIImage, didFinishSavingWithError error: Error?, contextInfo: UnsafeRawPointer) {
        if error != nil {
            showAnimatedAlertView(nil, message: "Save error")
        } else {
            showAnimatedAlertView(nil, message: "Saved!")
        }
    }
}
